src/core/HttpResponse.ts

```diff
@@ -0,0 +1,116 @@
+import type { DefaultBodyType } from './handlers/RequestHandler'
+import {
+  decorateResponse,
+  normalizeResponseInit,
+} from './utils/HttpResponse/decorators'
+
+export interface HttpResponseInit extends ResponseInit {
+  type?: ResponseType
+}
+
+declare const bodyType: unique symbol
+
+export interface StrictRequest<BodyType extends DefaultBodyType>
+  extends Request {
+  json(): Promise<BodyType>
+}
+
+/**
+ * Opaque `Response` type that supports strict body type.
+ */
+export interface StrictResponse<BodyType extends DefaultBodyType>
+  extends Response {
+  readonly [bodyType]: BodyType
+}
+
+/**
+ * A `Response` class superset with a stricter response body type.
+ * @example
+ * new HttpResponse('Hello world', { status: 201 })
+ */
+export class HttpResponse extends Response {
+  constructor(body?: BodyInit | null, init?: HttpResponseInit) {
+    const responseInit = normalizeResponseInit(init)
+    super(body, responseInit)
+    decorateResponse(this, responseInit)
+  }
+
+  /**
+   * Create a `Response` with a `Content-Type: "text/plain"` body.
+   * @example
+   * HttpResponse.text('hello world')
+   * HttpResponse.text('Error', { status: 500 })
+   */
+  static text<BodyType extends string>(
+    body?: BodyType | null,
+    init?: HttpResponseInit,
+  ): StrictResponse<BodyType> {
+    const responseInit = normalizeResponseInit(init)
+    responseInit.headers.set('Content-Type', 'text/plain')
+    return new HttpResponse(body, responseInit) as StrictResponse<BodyType>
+  }
+
+  /**
+   * Create a `Response` with a `Content-Type: "application/json"` body.
+   * @example
+   * HttpResponse.json({ firstName: 'John' })
+   * HttpResponse.json({ error: 'Not Authorized' }, { status: 401 })
+   */
+  static json<BodyType extends DefaultBodyType>(
+    body?: BodyType | null,
+    init?: HttpResponseInit,
+  ): StrictResponse<BodyType> {
+    const responseInit = normalizeResponseInit(init)
+    responseInit.headers.set('Content-Type', 'application/json')
```

Comment:
```suggestion
return Response.json(body, responseInit) as StrictResponse<BodyType>
```

I think this could be simplified slightly, but just ever so slightly

Reply:
very not necessry, but kinda cool

Reply:
It looks like `Response.json()` doesn't set the `Content-Length` header on the response. I think it'd be a great addition of `HttpResponse.json()` to set that header to always produce a complete and valid JSON response, the same way we should, if we don't already, set the `Content-Length` in other methods like `HttpResponse.blob()`.